### PR TITLE
RPP Tensor Water Augmentation on HOST and HIP

### PR DESCRIPTION
**Pull request by r-abishek**
@rrawther @asalmanp @kiritigowda

- Add support for RPP Tensor Water Augmentation with AVX2 on HOST and vectorization on HIP
- Adds respective U8 , F32 , I8 , F16 variants, PKD3, PLN3, PLN1 with toggle variants
- Adds test case forWater HOST and HIP kernels in new test suite, with Golden outputs for QA testing.
- Adds respective QA tests, unit tests and performance tests.

### CODE REVIEW

File: include/rppt_tensor_effects_augmentations.h [hunk 1]
```diff
@@ -164,6 +164,29 @@ RppStatus rppt_non_linear_blend_host(RppPtr_t srcPtr1, RppPtr_t srcPtr2, RpptDes
 RppStatus rppt_non_linear_blend_gpu(RppPtr_t srcPtr1, RppPtr_t srcPtr2, RpptDescPtr srcDescPtr, RppPtr_t dstPtr, RpptDescPtr dstDescPtr, Rpp32f *stdDevTensor, RpptROIPtr roiTensorPtrSrc, RpptRoiType roiType, rppHandle_t rppHandle);
 #endif // GPU_SUPPORT
 
+/******************** water ********************/
```
Comment (rrawther) [inline include/rppt_tensor_effects_augmentations.h:167]: please follow google coding spec for comment module. For example below
/*! \brief [Graph] Creates a Batch Normalization Layer Node.
 * \param [in] graph The handle to the graph.
 * \param [in] inputs The input tensor data.
 * \param [in] inputs The mean tensor data.
 * \param [in] inputs The variance tensor data.
 * \param [in] inputs The scale tensor data.
 * \param [in] inputs The bias tensor data.
 * \param [in] inputs The eps vx_float32 data.
 * \param [out] outputs The output tensor data.
 * \return <tt> vx_node</tt>.
 * \returns A node reference <tt>\ref vx_node</tt>. Any possible errors preventing a
 * successful creation should be checked using <tt>\ref vxGetStatus</tt>.
 */

File: src/include/cpu/rpp_cpu_simd.hpp [hunk 2]
```diff
@@ -2230,6 +2348,18 @@ inline void rpp_generic_nn_load_i8pln1(Rpp8s *srcPtrChannel, Rpp32s *srcLoc, Rpp
     p = _mm_unpacklo_epi8(px[0], px[1]);    // unpack to obtain [R01|R11|R21|R31|00|00|00|00|00|00|00|00|00|00|00|00]
 }
 
+inline void rpp_generic_nn_load_i8pln1_avx(Rpp8s *srcPtrChannel, Rpp32s *srcLoc, Rpp32s *invalidLoad, __m256i &p)
+{
+    Rpp8s buffer[16] = {0};
+    for(int i = 0; i < 8; i++)
+    {
```
Comment (rrawther) [inline src/include/cpu/rpp_cpu_simd.hpp:2355]: move the bracket to previous line. please follow google coding style

File: src/include/cpu/rpp_cpu_simd.hpp [hunk 3]
```diff
@@ -3101,16 +3231,74 @@ inline void rpp_resize_nn_load_i8pln1(Rpp8s *srcRowPtrsForInterp, Rpp32s *loc, _
 
 inline void rpp_store12_u8pkd3_to_u8pln3(Rpp8u* dstPtrR, Rpp8u* dstPtrG, Rpp8u* dstPtrB, __m128i &p)
 {
-    _mm_storeu_si128((__m128i *)(dstPtrR), _mm_shuffle_epi8(p, xmm_char_maskR)); /* Shuffle and extract the R pixels*/
-    _mm_storeu_si128((__m128i *)(dstPtrG), _mm_shuffle_epi8(p, xmm_char_maskG)); /* Shuffle and extract the G pixels*/
-    _mm_storeu_si128((__m128i *)(dstPtrB), _mm_shuffle_epi8(p, xmm_char_maskB)); /* Shuffle and extract the B pixels*/
+    _mm_storeu_si32((__m128i *)(dstPtrR), _mm_shuffle_epi8(p, xmm_char_maskR)); /* Shuffle and extract the R pixels*/
```
Comment (rrawther) [inline src/include/cpu/rpp_cpu_simd.hpp:3234]: please don't add spacing from previous (this has to be done as separate PR). When reviewing, it shows as difference in code and it is hard to review. It will be best to omit those changes

File: src/include/cpu/rpp_cpu_simd.hpp [hunk 4]
```diff
@@ -3139,6 +3343,22 @@ inline void rpp_store12_i8pln3_to_i8pkd3(Rpp8s* dstPtr, __m128i *p)
...
+inline void rpp_store24_i8pln3_to_i8pkd3_avx(Rpp8s* dstPtr, __m256i *p)
+{
+    __m128i px[5];
+    px[0] = _mm256_castsi256_si128(p[0]);            // [R01|R11|R21|R31|R41|R51|R61|R71|00|00|00|00|00|00|00|00]
+    px[1] = _mm256_castsi256_si128(p[1]);            // [G01|G11|G21|G31|G41|G51|G61|G71|00|00|00|00|00|00|00|00]
+    px[2] = _mm256_castsi256_si128(p[2]);            // [B01|B11|B21|B31|B41|B51|B61|B71|00|00|00|00|00|00|00|00]
+
+    px[3] = _mm_unpacklo_epi8(px[0], px[1]);         // [R01|G01|R11|G11|R21|G21|R31|G31|R41|G41|R51|G51|R61|G61|R71|G71]
+    px[4] = _mm_unpacklo_epi64(px[3], px[2]);        // [R01|G01|R11|G11|R21|G21|R31|G31|B01|B11|B21|B31|B41|B51|B61|B71]
+    _mm_storeu_si128((__m128i *)(dstPtr), _mm_shuffle_epi8(px[4], xmm_store4_pkd_pixels)); // shuffle to get RGB 00-03
```
Comment (rrawther) [inline src/include/cpu/rpp_cpu_simd.hpp:3355]: why are we not using _mm256_storeu_xxx instead?

**Review by rrawther — CHANGES_REQUESTED** [review]
We need to have a code review session to go through this code. It is hard to review intrinsics without knowing what the function is trying to do

File: src/include/cpu/rpp_cpu_simd.hpp (hunk 3 above)
Comment (sampath1117) [reply src/include/cpu/rpp_cpu_simd.hpp:3234]: Hi Rajy,
Here we changed from _mm_storeu_si128 to _mm_storeu_si32, since we want to store only 4 values
This is not a space change

File: src/include/cpu/rpp_cpu_simd.hpp (hunk 2 above)
Comment (sampath1117) [reply src/include/cpu/rpp_cpu_simd.hpp:2355]: Hi Rajy,
Currently all RPP functions follow the old RPP style where { will be in new line for functions
We planned to issue one global PR which will change all the instances to google style
Please let me know if you want this change to be done for water augmentation alone for now, and issue a global PR to change the style for the rest of the augmentations

File: src/include/cpu/rpp_cpu_simd.hpp [hunk 5]
```diff
@@ -2230,6 +2348,18 @@ inline void rpp_generic_nn_load_i8pln1(Rpp8s *srcPtrChannel, Rpp32s *srcLoc, Rpp
     p = _mm_unpacklo_epi8(px[0], px[1]);    // unpack to obtain [R01|R11|R21|R31|00|00|00|00|00|00|00|00|00|00|00|00]
 }
 
+inline void rpp_generic_nn_load_i8pln1_avx(Rpp8s *srcPtrChannel, Rpp32s *srcLoc, Rpp32s *invalidLoad, __m256i &p)
+{
+    Rpp8s buffer[16] = {0};
+    for(int i = 0; i < 8; i++)
+    {
+        if(!invalidLoad[i])
```
[the remark this answers is not in the document]
Comment (sampath1117) [reply src/include/cpu/rpp_cpu_simd.hpp:2356]: Hi Rajy,
I will try this out and get back to you on this

File: src/include/cpu/rpp_cpu_simd.hpp (hunk 4 above)
Comment (sampath1117) [reply src/include/cpu/rpp_cpu_simd.hpp:3355]: Hi Rajy,
Here we are doing the layout toggle from PLN3 to PKD3
we are extracting 128 bit from 256 bit register and using unpack instructions in SSE to get the data in PKD3 format

So we have PKD3 format data in SSE registers, we are using _mm_storeu_xx instead of _mm256_storeu_xxx

**Comment by rrawther** [comment]
@paveltc : Can you please run tests on this PR?

File: src/include/cpu/rpp_cpu_simd.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (sampath1117) [reply src/include/cpu/rpp_cpu_simd.hpp:2354]: Hi Rajy,
Here based on the mask value, we need to load from srcLoc[i] for a given pixel and this can vary for each pixel
_mm256_mask_i32gather_ps does precisely the same operation, but the counterpart is not available for U8 and I8 datatype

I have tried using _mm256_mask_i32gather_epi32 + permute to rearrange the data in required order for U8, but it is degrading the performance. so i have retained the existing code without making any changes

**Comment by kiritigowda** [comment]
@r-abishek is this ready to be merged? @paveltc

File: utilities/test_suite/HOST/testAllScript.sh [hunk 6]
```diff
@@ -4,7 +4,7 @@ cwd=$(pwd)
```
Comment (rrawther) [inline utilities/test_suite/HOST/testAllScript.sh:4]: are the csv files supposed to be checked in as well?

**Review by rrawther — APPROVED** [review]
@asalmanp to review HIP kernels
@paveltc to run unit tests

File: utilities/test_suite/HOST/testAllScript.sh (hunk 6 above)
Comment (r-abishek) [reply utilities/test_suite/HOST/testAllScript.sh:4]: The CSV basically contains the golden ouptut for small 50x50/100x100/150x150 images to pass QA tests

**Comment by paveltc** [comment]
@kiritigowda @rrawther This PR passes all relevant unit tests.

**Comment by kiritigowda** [comment]
@r-abishek can you fix the merge conflicts?

**Comment by r-abishek** [comment]
> @r-abishek can you fix the merge conflicts?

@kiritigowda resolved

**Comment by kiritigowda** [comment]
@r-abishek failing compilation 

```
In file included from /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/modules/rppi_advanced_augmentations.cpp:26:
In file included from /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/modules/cpu/host_advanced_augmentations.hpp:26:
/var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/include/cpu/rpp_cpu_simd.hpp:3251:5: error: use of undeclared identifier '_mm_storeu_si32'; did you mean '_mm_storeu_si128'?
    _mm_storeu_si32((__m128i *)(dstPtrG + 4), _mm_shuffle_epi8(p128[1], xmm_char_maskG)); /* shuffle to get G05-G08*/
    ^~~~~~~~~~~~~~~
    _mm_storeu_si128
/opt/rh/llvm-toolset-7/root/usr/lib64/clang/5.0.1/include/emmintrin.h:4001:1: note: '_mm_storeu_si128' declared here
_mm_storeu_si128(__m128i *__p, __m128i __b)
^
In file included from /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/modules/rppi_advanced_augmentations.cpp:26:
In file included from /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/modules/cpu/host_advanced_augmentations.hpp:26:
/var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/include/cpu/rpp_cpu_simd.hpp:3252:5: error: use of undeclared identifier '_mm_storeu_si32'; did you mean '_mm_storeu_si128'?
    _mm_storeu_si32((__m128i *)(dstPtrB + 4), _mm_shuffle_epi8(p128[1], xmm_char_maskB)); /* shuffle to get B05-B08*/
    ^~~~~~~~~~~~~~~
    _mm_storeu_si128
/opt/rh/llvm-toolset-7/root/usr/lib64/clang/5.0.1/include/emmintrin.h:4001:1: note: '_mm_storeu_si128' declared here
_mm_storeu_si128(__m128i *__p, __m128i __b)

```

View summary - http://math-ci.amd.com/blue/organizations/jenkins/main%2FShort-GPUOpen%2Frpp/detail/PR-181/9/pipeline/472

**Comment by r-abishek** [comment]
> @r-abishek failing compilation on CentOS
> 
> ```
> In file included from /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/modules/rppi_advanced_augmentations.cpp:26:
> In file included from /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/modules/cpu/host_advanced_augmentations.hpp:26:
> /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/include/cpu/rpp_cpu_simd.hpp:3251:5: error: use of undeclared identifier '_mm_storeu_si32'; did you mean '_mm_storeu_si128'?
>     _mm_storeu_si32((__m128i *)(dstPtrG + 4), _mm_shuffle_epi8(p128[1], xmm_char_maskG)); /* shuffle to get G05-G08*/
>     ^~~~~~~~~~~~~~~
>     _mm_storeu_si128
> /opt/rh/llvm-toolset-7/root/usr/lib64/clang/5.0.1/include/emmintrin.h:4001:1: note: '_mm_storeu_si128' declared here
> _mm_storeu_si128(__m128i *__p, __m128i __b)
> ^
> In file included from /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/modules/rppi_advanced_augmentations.cpp:26:
> In file included from /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/modules/cpu/host_advanced_augmentations.hpp:26:
> /var/jenkins_home/workspace/main_Short-GPUOpen_rpp_PR-181/rbUpkl7HK/rpp/src/include/cpu/rpp_cpu_simd.hpp:3252:5: error: use of undeclared identifier '_mm_storeu_si32'; did you mean '_mm_storeu_si128'?
>     _mm_storeu_si32((__m128i *)(dstPtrB + 4), _mm_shuffle_epi8(p128[1], xmm_char_maskB)); /* shuffle to get B05-B08*/
>     ^~~~~~~~~~~~~~~
>     _mm_storeu_si128
> /opt/rh/llvm-toolset-7/root/usr/lib64/clang/5.0.1/include/emmintrin.h:4001:1: note: '_mm_storeu_si128' declared here
> _mm_storeu_si128(__m128i *__p, __m128i __b)
> ```
> 
> View summary - http://math-ci.amd.com/blue/organizations/jenkins/main%2FShort-GPUOpen%2Frpp/detail/PR-181/9/pipeline/472

@kiritigowda _mm_storeu_si32() seems to be defined in emmintrin.h (screenshot below). 
Not sure how Ubuntu is able to pick it up but not CentOS.
Also, we don't have an assigned CentOS machine to check this. Seeing this kind of issue for the first time.

image

**Comment by kiritigowda** [comment]
@r-abishek let me look into this issue and see if this could be a fixed during setup or #def code change

**Comment by r-abishek** [comment]
@kiritigowda Would it be possible to issue a separate PR for this fix and merge #181 , #206 and #179 since we need these changes for some next upcoming PRs?

**Comment by kiritigowda** [comment]
@r-abishek can you look into the conflicts?

**Comment by r-abishek** [comment]
@kiritigowda Conflicts resolved

**Review by rrawther — APPROVED** [review]
@asalmanp Can you please review the HIP kernels for this PR?
@paveltc to run unit tests

**Comment by kiritigowda** [comment]
@paveltc let me know if this branch is good to merge?

**Comment by paveltc** [comment]
@kiritigowda This can be merged.

**Comment by r-abishek** [comment]
@kiritigowda @paveltc Yes, this could be merged except for that one issue - with the CentOS build.
We have tried pulling the docker used by the CI with VPN. This is what it uses - 10.216.51.87:5000/rocm-plus-docker/compute-rocm-dkms-no-npi-hipclang:12755-centos-7.x-stg1

But are getting either of the following issues:
Error response from daemon: Get "[https://10.216.51.87:5000/v2/":](https://10.216.51.87:5000/v2/%22:) http: server gave HTTP response to HTTPS client
Error response from daemon: pull access denied for compute-rocm-dkms-no-npi-hipclang, repository does not exist or may require 'docker login': denied: requested access to the resource is denied

**Comment by kiritigowda** [comment]
@r-abishek - you need to create the below file to pull that docker.

 sudo vi /etc/docker/daemon.json

```
{
  "insecure-registries" : ["compute-artifactory.amd.com:5000"]
}
```

**Comment by kiritigowda** [comment]
@r-abishek also there is a conflict in this PR.
